src/security/KeyData.cc

```diff
@@ -94,51 +133,40 @@ Security::KeyData::loadX509ChainFromFile()
-
-        while (auto ca = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr)) {
-            // get Issuer name of the cert for debug display
-            char *nameStr = X509_NAME_oneline(X509_get_subject_name(ca), nullptr, 0);
-
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
-            // self-signed certificates are not valid in a sent chain
-            if (X509_check_issued(ca, ca) == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
-                OPENSSL_free(nameStr);
-                continue;
-            }
-#endif
-            // checks that the chained certs are actually part of a chain for validating cert
-            const auto checkCode = X509_check_issued(ca, latestCert.get());
-            if (checkCode == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
-                // OpenSSL API requires that we order certificates such that the
-                // chain can be appended directly into the on-wire traffic.
-                latestCert = CertPointer(ca);
-                chain.emplace_front(latestCert);
-            } else {
-                debugs(83, DBG_PARSE_NOTE(2), certFile << ": Ignoring non-issuer CA " << nameStr << ": " << X509_verify_cert_error_string(checkCode) << " (" << checkCode << ")");
-            }
-            OPENSSL_free(nameStr);
-        }
+    while (CertPointer ca = CertPointer(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr))) {
+        verifyChainCert(ca, latestCert, chain);
     }
 
 #elif USE_GNUTLS
-    // XXX: implement chain loading
-    debugs(83, 2, "Loading certificate chain from PEM files not implemented in this Squid.");
+    const char *certFilename = certFile.c_str();
+    gnutls_datum_t data;
+    Security::ErrorCode x = gnutls_load_file(certFilename, &data);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to load chain file '" << certFile << "': " << ErrorString(x));
+        return;
+    }
+
+    unsigned int listSz = 0;
+    gnutls_x509_crt_t *certChain;
+    x = gnutls_x509_crt_list_import2(&certChain, &listSz, &data, GNUTLS_X509_FMT_PEM, 0);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to import chain file '" << certFile << "': " << ErrorString(x));
+        return;
+    }
+
+    for (unsigned int i = 0; i < listSz ; ++i) {
+        CertPointer ca = Security::CertPointer(certChain[i], [](gnutls_x509_crt_t p) {
+            debugs(83, 5, "gnutls_x509_crt_deinit cert=" << (void*)p);
+            gnutls_x509_crt_deinit(p);
+        });
+
+        verifyChainCert(ca, latestCert, chain);
+    }
+    gnutls_free(certChain);
 
 #else
-    // nothing to do.
+    auto nameStr = CertSubjectName(cert);
+    debugs(83, DBG_PARSE_NOTE(2), "Loading certificate chain of " << nameStr << " from PEM files not implemented in this Squid.");
```

Comment:
Lacking chain load does not imply lack of cert details. As you know right now GnuTLS already has cert support, but not chain load support. Done anyway since it does not matter now until the next library is added.